src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderImpl.cpp

```diff
@@ -0,0 +1,639 @@
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    member = it->second;
+    return RETCODE_OK;
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_all_members(
+        DynamicTypeMembersById& member) noexcept
+{
+    member = member_;
+    return RETCODE_OK;
+}
+
+uint32_t DynamicTypeBuilderImpl::get_member_count() noexcept
+{
+    return static_cast<uint32_t>(members_.size());
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_member_by_index(
+        traits<DynamicTypeMember>::ref_type& member,
+        uint32_t index) noexcept
+{
+    if (index >= members_.size())
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Index " << index << " out-of-range");
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    member = members_.at(index);
+    return RETCODE_OK;
+}
+
+uint32_t DynamicTypeBuilderImpl::get_annotation_count() noexcept
+{
+    return static_cast<uint32_t>(annotation_.size());
+}
+
+ReturnCode_t DynamicTypeBuilderImpl::get_annotation(
+        traits<AnnotationDescriptor>::ref_type& descriptor,
+        uint32_t idx) noexcept
+{
+    if (!descriptor || idx >= annotation_.size())
+    {
+        EPROSIMA_LOG_ERROR(DYN_TYPES, "Descriptor reference is nil or index is out-of-range");
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    traits<AnnotationDescriptor>::narrow<AnnotationDescriptorImpl>(descriptor)->copy_from(annotation_.at(idx));
+    return RETCODE_OK;
+}
+
+bool DynamicTypeBuilderImpl::equals(
+        traits<DynamicType>::ref_type other) noexcept
+{
+    bool ret_value = true;
+    auto impl = traits<DynamicType>::narrow<DynamicTypeImpl>(other);
+
+    ret_value &= type_descriptor_.equals(impl->type_descriptor_);
+    if (ret_value)
+    {
+        ret_value &= annotation_.size() == impl->annotation_.size();
+        if (ret_value)
+        {
+            for (size_t count {0}; ret_value && count < annotation_.size(); ++count)
+            {
+                ret_value &= annotation_.at(count).equals(impl->annotation_.at(count));
+            }
+        }
+
+        ret_value &= member_.size() == impl->member_.size();
+        assert(TK_ANNOTATION == type_descriptor_.kind() ||
+                TK_STRUCTURE == type_descriptor_.kind() ||
+                TK_UNION ==  type_descriptor_.kind() ||
+                0 == member_.size());
+        assert(TK_ANNOTATION == type_descriptor_.kind() ||
+                TK_STRUCTURE == impl->type_descriptor_.kind() ||
+                TK_UNION ==  impl->type_descriptor_.kind() ||
+                0 == impl->member_.size());
+        assert((TK_ANNOTATION != type_descriptor_.kind() &&
+                TK_STRUCTURE != type_descriptor_.kind() &&
+                TK_UNION != type_descriptor_.kind()) ||
+                0 < member_.size());
+        assert((TK_ANNOTATION != type_descriptor_.kind() &&
+                TK_STRUCTURE != impl->type_descriptor_.kind() &&
+                TK_UNION != impl->type_descriptor_.kind()) ||
+                0 < member_.size());
+
```

Comment:
On the one hand, having just checked that member_.size() is equal to impl_.size() and `type_descriptor_` equals `impl->type_descriptor`, I think it is enough to assert one of the tow but not both of them. In fact the second and fourth assertion are mixing information coming from the `other` variable and from `this`. Thus I think it is enough with the first assertion, as the third one is already checked in the first one.

Reply:
Done in 8f6d0ffb4